File: ios/LocationStore.swift
import Foundation
import SQLite3

@objc public class LocationStore: NSObject {
  @objc public static let shared = LocationStore()

  private let queue = DispatchQueue(label: "com.echobitzit.backgroundlocation.locationstore")
  private var db: OpaquePointer?
  private let maxRowsPerSession = 10_000

  private override init() {
    super.init()
    openDatabase()
    createTableIfNeeded()
  }

  deinit {
    if db != nil {
      sqlite3_close(db)
    }
  }

  @objc public func insert(
    sessionId: String,
    latitude: Double,
    longitude: Double,
    accuracy: Double?,
    speed: Double?,
    heading: Double?,
    altitude: Double?,
    timestamp: Double
  ) {
    queue.sync {
      guard let db = db else { return }
      let sql = """
      INSERT INTO locations (
        session_id, latitude, longitude, accuracy, speed, heading, altitude, timestamp
      ) VALUES (?, ?, ?, ?, ?, ?, ?, ?)
      """
      var statement: OpaquePointer?
      guard sqlite3_prepare_v2(db, sql, -1, &statement, nil) == SQLITE_OK else {
        return
      }
      defer { sqlite3_finalize(statement) }

      sqlite3_bind_text(statement, 1, (sessionId as NSString).utf8String, -1, nil)
      sqlite3_bind_double(statement, 2, latitude)
      sqlite3_bind_double(statement, 3, longitude)
      if let accuracy = accuracy {
        sqlite3_bind_double(statement, 4, accuracy)
      } else {
        sqlite3_bind_null(statement, 4)
      }
      if let speed = speed {
        sqlite3_bind_double(statement, 5, speed)
      } else {
        sqlite3_bind_null(statement, 5)
      }
      if let heading = heading {
        sqlite3_bind_double(statement, 6, heading)
      } else {
        sqlite3_bind_null(statement, 6)
      }
      if let altitude = altitude {
        sqlite3_bind_double(statement, 7, altitude)
      } else {
        sqlite3_bind_null(statement, 7)
      }
      sqlite3_bind_double(statement, 8, timestamp)

      if sqlite3_step(statement) != SQLITE_DONE {
        print("[EBBgLoc] LocationStore insert failed")
      }
      trimSession(sessionId: sessionId)
    }
  }

  @objc public func query(sessionId: String, limit: Int) -> [[String: Any]] {
    queue.sync {
      guard let db = db else { return [] }
      let safeLimit = limit > 0 ? limit : 500
      let sql = """
      SELECT session_id, latitude, longitude, accuracy, speed, heading, altitude, timestamp
      FROM locations
      WHERE session_id = ?
      ORDER BY timestamp DESC
      LIMIT ?
      """
      var statement: OpaquePointer?
      guard sqlite3_prepare_v2(db, sql, -1, &statement, nil) == SQLITE_OK else {
        return []
      }
      defer { sqlite3_finalize(statement) }

      sqlite3_bind_text(statement, 1, (sessionId as NSString).utf8String, -1, nil)
      sqlite3_bind_int(statement, 2, Int32(safeLimit))

      var results: [[String: Any]] = []
      while sqlite3_step(statement) == SQLITE_ROW {
        let sid = String(cString: sqlite3_column_text(statement, 0))
        let latitude = sqlite3_column_double(statement, 1)
        let longitude = sqlite3_column_double(statement, 2)
        var row: [String: Any] = [
          "sessionId": sid,
          "latitude": latitude,
          "longitude": longitude,
          "timestamp": sqlite3_column_double(statement, 7),
        ]
        if sqlite3_column_type(statement, 3) != SQLITE_NULL {
          row["accuracy"] = sqlite3_column_double(statement, 3)
        }
        if sqlite3_column_type(statement, 4) != SQLITE_NULL {
          row["speed"] = sqlite3_column_double(statement, 4)
        }
        if sqlite3_column_type(statement, 5) != SQLITE_NULL {
          row["heading"] = sqlite3_column_double(statement, 5)
        }
        if sqlite3_column_type(statement, 6) != SQLITE_NULL {
          row["altitude"] = sqlite3_column_double(statement, 6)
        }
        results.append(row)
      }
      return results
    }
  }

  @objc public func clear(sessionId: String?) {
    queue.sync {
      guard let db = db else { return }
      if let sessionId = sessionId, !sessionId.isEmpty {
        let sql = "DELETE FROM locations WHERE session_id = ?"
        var statement: OpaquePointer?
        guard sqlite3_prepare_v2(db, sql, -1, &statement, nil) == SQLITE_OK else { return }
        defer { sqlite3_finalize(statement) }
        sqlite3_bind_text(statement, 1, (sessionId as NSString).utf8String, -1, nil)
        sqlite3_step(statement)
        print("[EBBgLoc] LocationStore clear session=\(sessionId)")
      } else {
        sqlite3_exec(db, "DELETE FROM locations", nil, nil, nil)
        print("[EBBgLoc] LocationStore clear all")
      }
    }
  }

  private func openDatabase() {
    let urls = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)
    let base = urls.first ?? FileManager.default.temporaryDirectory
    let dir = base.appendingPathComponent("eb_background_location", isDirectory: true)
    try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
    let path = dir.appendingPathComponent("locations.db").path
    if sqlite3_open(path, &db) != SQLITE_OK {
      print("[EBBgLoc] LocationStore failed to open db at \(path)")
      db = nil
    }
  }

  private func createTableIfNeeded() {
    guard let db = db else { return }
    let sql = """
    CREATE TABLE IF NOT EXISTS locations (
      id INTEGER PRIMARY KEY AUTOINCREMENT,
      session_id TEXT NOT NULL,
      latitude REAL NOT NULL,
      longitude REAL NOT NULL,
      accuracy REAL,
      speed REAL,
      heading REAL,
      altitude REAL,
      timestamp REAL NOT NULL
    );
    CREATE INDEX IF NOT EXISTS idx_locations_session_timestamp
      ON locations(session_id, timestamp DESC);
    """
    sqlite3_exec(db, sql, nil, nil, nil)
  }

  private func trimSession(sessionId: String) {
    guard let db = db else { return }
    let sql = """
    DELETE FROM locations
    WHERE session_id = ?
      AND id NOT IN (
        SELECT id FROM locations
        WHERE session_id = ?
        ORDER BY timestamp DESC
        LIMIT ?
      )
    """
    var statement: OpaquePointer?
    guard sqlite3_prepare_v2(db, sql, -1, &statement, nil) == SQLITE_OK else { return }
    defer { sqlite3_finalize(statement) }
    sqlite3_bind_text(statement, 1, (sessionId as NSString).utf8String, -1, nil)
    sqlite3_bind_text(statement, 2, (sessionId as NSString).utf8String, -1, nil)
    sqlite3_bind_int(statement, 3, Int32(maxRowsPerSession))
    sqlite3_step(statement)
  }
}

File: android/src/main/java/com/echobitzit/backgroundlocation/LocationStore.kt
package com.echobitzit.backgroundlocation

import android.content.Context
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import android.util.Log

/**
 * SQLite persistence for session location points.
 */
class LocationStore(context: Context) {
  private val dbHelper = LocationDbHelper(context.applicationContext)

  fun insert(
    sessionId: String,
    latitude: Double,
    longitude: Double,
    accuracy: Float?,
    speed: Float?,
    heading: Float?,
    altitude: Double?,
    timestamp: Long,
  ) {
    val db = dbHelper.writableDatabase
    val values = android.content.ContentValues().apply {
      put("session_id", sessionId)
      put("latitude", latitude)
      put("longitude", longitude)
      if (accuracy != null) put("accuracy", accuracy.toDouble())
      if (speed != null) put("speed", speed.toDouble())
      if (heading != null) put("heading", heading.toDouble())
      if (altitude != null) put("altitude", altitude)
      put("timestamp", timestamp)
    }
    db.insert("locations", null, values)
    trimSession(sessionId)
  }

  fun query(sessionId: String, limit: Int): List<StoredLocation> {
    val db = dbHelper.readableDatabase
    val safeLimit = if (limit > 0) limit else DEFAULT_QUERY_LIMIT
    val cursor = db.rawQuery(
      """
      SELECT session_id, latitude, longitude, accuracy, speed, heading, altitude, timestamp
      FROM locations
      WHERE session_id = ?
      ORDER BY timestamp DESC
      LIMIT ?
      """.trimIndent(),
      arrayOf(sessionId, safeLimit.toString()),
    )
    val results = mutableListOf<StoredLocation>()
    cursor.use {
      while (it.moveToNext()) {
        results.add(
          StoredLocation(
            sessionId = it.getString(0),
            latitude = it.getDouble(1),
            longitude = it.getDouble(2),
            accuracy = if (it.isNull(3)) null else it.getFloat(3),
            speed = if (it.isNull(4)) null else it.getFloat(4),
            heading = if (it.isNull(5)) null else it.getFloat(5),
            altitude = if (it.isNull(6)) null else it.getDouble(6),
            timestamp = it.getLong(7),
          )
        )
      }
    }
    return results
  }

  fun clear(sessionId: String?) {
    val db = dbHelper.writableDatabase
    if (sessionId.isNullOrBlank()) {
      db.delete("locations", null, null)
      Log.d(TAG, "clear all session locations")
    } else {
      db.delete("locations", "session_id = ?", arrayOf(sessionId))
      Log.d(TAG, "clear session locations session=$sessionId")
    }
  }

  private fun trimSession(sessionId: String) {
    val db = dbHelper.writableDatabase
    db.execSQL(
      """
      DELETE FROM locations
      WHERE session_id = ?
        AND id NOT IN (
          SELECT id FROM locations
          WHERE session_id = ?
          ORDER BY timestamp DESC
          LIMIT ?
        )
      """.trimIndent(),
      arrayOf(sessionId, sessionId, MAX_ROWS_PER_SESSION.toString()),
    )
  }

  data class StoredLocation(
    val sessionId: String,
    val latitude: Double,
    val longitude: Double,
    val accuracy: Float?,
    val speed: Float?,
    val heading: Float?,
    val altitude: Double?,
    val timestamp: Long,
  )

  private class LocationDbHelper(context: Context) :
    SQLiteOpenHelper(context, DB_NAME, null, DB_VERSION) {
    override fun onCreate(db: SQLiteDatabase) {
      db.execSQL(
        """
        CREATE TABLE locations (
          id INTEGER PRIMARY KEY AUTOINCREMENT,
          session_id TEXT NOT NULL,
          latitude REAL NOT NULL,
          longitude REAL NOT NULL,
          accuracy REAL,
          speed REAL,
          heading REAL,
          altitude REAL,
          timestamp INTEGER NOT NULL
        )
        """.trimIndent()
      )
      db.execSQL(
        "CREATE INDEX idx_locations_session_timestamp ON locations(session_id, timestamp DESC)"
      )
    }

    override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {
      db.execSQL("DROP TABLE IF EXISTS locations")
      onCreate(db)
    }
  }

  companion object {
    private const val TAG = "EBBgLoc"
    private const val DB_NAME = "eb_background_locations.db"
    private const val DB_VERSION = 1
    private const val MAX_ROWS_PER_SESSION = 10_000
    private const val DEFAULT_QUERY_LIMIT = 500
  }
}
